src/lib/common/mtmodule.py

```diff
@@ -1,17 +1,71 @@
 from abc import ABC, abstractmethod
-from lib.common.util import save_logs
-from lib.common.exceptions import ImproperLoggedPhaseError
+from lib.common.util import save_logs, hashdict
+from lib.common.exceptions import ImproperLoggedPhaseError, BatchedPhaseArgNotGenerator
 from lib.common.etypes import Etype
 from functools import partial, wraps
+from types import GeneratorType
+from itertools import islice, chain
 import os
+import multiprocessing
+import struct
 
+MAX_CPUS = multiprocessing.cpu_count() - 1
+MIN_ELEMENTS_PER_CPU = 4
+
+
+def get_batch_size(ls_len):
+    """ Determine the batch size for multiprocessing. """
+    if ls_len > MAX_CPUS * MIN_ELEMENTS_PER_CPU:
+        return ls_len // MAX_CPUS + 1
+    # TODO: improve this heuristic for splitting up jobs
+    return ls_len
+
+
+def chunks(iterable, size=1):
+    iterator = iter(iterable)
+    for first in iterator:
+        yield chain([first], islice(iterator, size - 1))
+
+
+def batch(iterable, n=1):
+    l = len(iterable)
+    for ndx in range(0, l, n):
+        yield iterable[ndx : min(ndx + n, l)]
+
+def db_run(dbfile, q, batches_running):
+    with open(dbfile, 'ab') as f:
+        while batches_running.value is not 0:
+            try:
+                done_info = q.get_nowait()             
+                f.write(struct.pack('II', *done_info))
+                f.flush()
+            except:
+                pass
+        while q.qsize() > 0:   
+            done_info = q.get()             
+            f.write(struct.pack('II', *done_info))
+            f.flush()
+
+        f.close()
+
+def process_batch(innards, self, done_dict, done_queue, batch_num, c, other_args):
+    for idx, i in enumerate(c):
+        if idx not in done_dict:            
+            innards(self, [i], *other_args)
+            done_queue.put((batch_num, idx))
+        else:
+            print("Batch %d item %d already done, skipping job." % (batch_num, idx))
```

Comment:
These non-class functions could either go in 'util.py', or even a new 'batching.py' or something similar- if you're up for it?

Reply:
Sure. Also, I realized that I left that print in there which was just there for debugging. Should probably remove it or switch to logger.

Reply:
I've switched this to a class function, but tests fail when I change the print statement to `self.logger`. This might be because self.logger add lines to the generated .txt files, I'm not sure. Would be good to keep the print statement in there as a log if it's not too difficult.

Reply:
@breezykermo Ah, yeah the error is because the test is using number of log lines to test how many elements it's processed, so this additional log throws it off.